src/inmanta/server/agentmanager.py

```diff
@@ -621,11 +621,24 @@ def get_agent_client(self, tid: uuid.UUID, endpoint: str, live_agent_only: bool
         else:
             return None
 
-    def are_agents_up(self, tid: uuid.UUID, endpoints: List[str]) -> bool:
-        return all([self.is_agent_up(tid, e) for e in endpoints])
+    async def are_agents_active(self, tid: uuid.UUID, endpoints: List[str]) -> bool:
+        """
+        Return true iff all the given agents are in the up or the paused state.
+        """
+        are_active: abc.Sequence[bool] = await asyncio.gather(*[self.is_agent_active(tid, e) for e in endpoints])
+        return all(are_active)
 
-    def is_agent_up(self, tid: uuid.UUID, endpoint: str) -> bool:
-        return (tid, endpoint) in self.tid_endpoint_to_session
+    async def is_agent_active(self, tid: uuid.UUID, endpoint: str) -> bool:
+        """
+        Return true iff the given agent is in the up or the paused state.
+        """
+        if (tid, endpoint) in self.tid_endpoint_to_session:
+            # Agent is up
+            return True
+        # Verify if there is an AgentInstance for the given endpoint.
+        # This case happens when the Agent is paused.
+        result = await data.AgentInstance.active_for(tid=tid, endpoint=endpoint)
```

Comment:
Why not use `endpoints_for_sid`?

This method mixes in memory with database structures, which can produce very complex behavior

Reply:
Mainly because there wasn't an in-memory data structure that was indexed on suitable value, but I see your concern. I will fix it.

Reply:
How would you use `endpoints_for_sid`? That data structure doesn't contain any information regarding the environment that the endpoint belongs to.